### ISSUE TITLE
research(system-one): evaluate Laya + PRAXIST training pipeline for #233

### ISSUE THREAD

**Issue by dyphn1** [dyphn1/Harness-everything#255]
## 目標

針對 #233 的 Harness System One router / intent scoring，建立一條獨立研究路徑，評估是否以 **Laya** 作為可訓練 typed-decision substrate，並以 **PRAXIST** 編排可重現的研究/實驗迭代。

本 issue 不直接取代 #233 已有的 runtime、fallback、structural floor 或 rollout gate；它的責任是產生可驗證的訓練方案、模型 artifact 與 evidence，供 #233 決定是否採用。

## 背景

#233 已完成/建立：
- owner-reviewed tier/intent holdout 與 graded agreement。
- intent 已從單一 primary/secondary 分類，朝 **per-intent relevance score (0..1)** 方向演進。
- deterministic structural floor / fallback / abstain / rollout gates。
- resident provider 與 CPU latency evidence。
- private training corpus 位於 `dyphn1/harness-system-one-data`，不得複製 prompt 至 public repo。

目前 n-gram 模型主要限制：
- tier3 依賴 scope / context，character n-gram 難以捕捉。
- intent model coverage 很低；hard/稀疏 primary-secondary labels 對多 intent relevance 的 supervision 不足。

## 研究假設

1. Laya 的 typed decisions / parallel decisions 可映射為 Harness 的 per-intent relevance scoring。
2. soft target（例如 `true=p, false=1-p`）可直接承接 #233 的 dense teacher scores。
3. active task / previous request 等 context 應納入 state，處理 `keep going` 等 continuation prompt。
4. tier 可由模型提供 proposal，但 **不得移除 #233 deterministic structural floor**。
5. PRAXIST 應負責 research orchestration / experiment selection / evidence aggregation，而不是成為 production runtime dependency。

## 非目標

- 不在本 issue 重新定義 #233 的 tier/intent taxonomy。
- 不把 owner-reviewed holdout 放進 training。
- 不把 private prompts 搬進 public repository。
- 不讓 learned model 取代 deterministic structural constraints。
- 不預設 Laya 一定優於現有 n-gram / CUA-S1；必須由 evidence 決定。
- 不把 PRAXIST 放進 production prompt path。

## 上游

- #233
- https://github.com/NandhaKishorM/laya
- https://github.com/sapientinc/PRAXIST

## 交付原則

每個 phase 一律依序：
1. 文件 / contract
2. 測試先 RED
3. 實作 GREEN
4. 產生可重現 evidence
5. commit / PR 保留 phase 邊界

詳細 phase、驗收方式與 stop/go gate 另以 comment 補上。

**Comment by dyphn1** [comment]
## Phase plan / implementation & acceptance

這個 issue 建議切成 **9 個 phase（0–8）**。核心原則是：**#255 只負責把研究問題變成可重現、可量測、可停止的工程流程；#233 仍是 production contract 的 single source of truth。**

---

### Phase 0 — Research contract / upstream pinning

**目的**

先固定研究邊界，避免 Laya / PRAXIST upstream 漂移、license、checkpoint 或 evaluator 變化，導致後續 evidence 無法重現。

**文件**
- 新增 #255 research contract。
- 固定：
  - Laya source revision / package version。
  - 使用 checkpoint 名稱、revision、artifact hash。
  - PRAXIST source/package revision。
  - Python / torch / transformers / ONNX Runtime（若使用）版本。
  - #233 evaluator revision。
  - private corpus commit。
- 明列 train / validation / research-dev / final-holdout 的用途。
- 明列禁止事項：final holdout 不可進 training、不可讓 research peer 修改 evaluator、不可把 private prompt 帶進 public artifact。

**RED tests**
- 未 pin revision/hash 時 readiness fail。
- evaluator / corpus revision 不符時 fail。
- final-holdout 出現在 train input 時 fail。
- public evidence bundle 含 raw private prompt 時 fail。

**GREEN implementation**
- machine-readable research manifest。
- readiness validator。
- evidence metadata 自動帶 provenance。

**驗收**
- baseline 可在 **不安裝 PRAXIST** 的情況下獨立執行。
- 同一 manifest 連跑兩次，輸入集合/hash/metric schema 一致。
- Laya zero-shot smoke test 可執行。
- PRAXIST 僅完成 doctor / task readiness，不開始正式研究。
- repository scan 證明沒有 private prompts 被寫入 public repo。

**STOP**
- license / provenance / checkpoint revision 無法固定時，不進 Phase 1。

---

### Phase 1 — Dense intent score data contract

**目的**

把 #233 現在的 sparse `primary + secondary[]` supervision，轉成 Laya 可學習的 **per-intent relevance score**：

```json
{
  "i": 123,
  "scores": {
    "fix": 0.92,
    "edit": 0.70,
    "test": 0.64,
    "investigate": 0.51
  }
}
```

未列出的 intent 必須由 schema 明確定義是 0、unknown 或 forbidden；不可讓不同 trainer 自行解讀。

**文件**
- score 定義與 band：
  - primary relevance
  - secondary relevance
  - related
  - irrelevant
- active-task context schema：
  - current prompt
  - previous actionable request
  - active task kind/scope/phase
- owner override / exclusion precedence。
- multilingual normalization policy。

**RED tests**
- score <0 或 >1 fail。
- intent key 缺漏/未知 fail。
- duplicate sample id fail。
- owner override 被 teacher label 蓋掉 fail。
- continuation sample 缺 active context 時標記 invalid / incomplete。
- final holdout 不能被 dense-label pipeline 改寫。

**GREEN implementation**
- dense label schema + validator。
- labeler/exporter。
- sparse → dense migration 只可作 bootstrap，不可冒充 owner truth。
- private corpus 保持 private。

**驗收**
- 至少建立一個 **獨立 research-dev review set**（不使用 final holdout）做人工 spot-check。
- en / zh-TW 都要有樣本。
- 每個 intent 都能產生統計：score distribution、band count、missing rate。
- owner override 100% preserved。
- private-data leakage = 0。

> 這一 phase 不以「模型準確率」驗收，而以 **label semantics 是否穩定且可審核** 驗收。若 review 顯示 score 定義本身仍有系統性歧義，回到文件修規則，不進 trainer。

---

### Phase 2 — Laya adapter + zero-shot baseline

**目的**

先證明 Harness 的 state/questions 可以穩定映射到 Laya，而不是直接開始 fine-tune。

**建議 mapping**
- intent：每 intent 一個 `noul`，取 yes probability。
- tier：可另外做 `choice` proposal，但 final tier 仍經 #233 deterministic floor。
- state：prompt-only 與 prompt+active-context 都保留為可比較模式。

**文件**
- Harness ↔ Laya request/response mapping。
- unavailable / timeout / invalid-output / model-mismatch 行為。
- question ordering 與 output normalization。

**RED tests**
- intent order 改變不應改變 semantic mapping。
- missing answer / NaN / malformed probability → invalid。
- bilingual fixtures。
- continuation fixtures（例如 `keep going`）。
- tier proposal 不得 bypass structural floor。

**GREEN implementation**
- Laya provider adapter。
- zero-shot evaluator path。
- 不接 production router，僅 benchmark。

**驗收**
- frozen fixtures 連跑結果可重現。
- 所有 12 intent 都有 0..1 score。
- failure 可 deterministic fallback。
- 產出 zero-shot evidence bundle，作為後續 trainer baseline。
- 即使 zero-shot 很差，只要 adapter contract 正確，此 phase 仍可通過；品質由後續 phase 判斷。

---

### Phase 3 — Fine-tuning pipeline / soft targets

**目的**

建立真正可重現的 Laya training pipeline，先不引入 PRAXIST。

**訓練 target**
對 intent relevance `p`：
- yes = `p`
- no = `1-p`

先跑至少兩條候選：
1. `laya-multilingual`
2. `laya-typed-decisions` / 其他 pinned upstream checkpoint

**文件**
- dataset conversion。
- split policy。
- seed policy。
- calibration policy。
- checkpoint naming / manifest / hash。
- GPU/CPU resource assumptions。

**RED tests**
- train/validation/final-holdout overlap fail。
- conversion 後 target distribution 不守恆 fail。
- fixed seed 下 sample ordering 必須一致。
- checkpoint 缺 manifest/hash fail。
- calibration 不得讀 final holdout。

**GREEN implementation**
- training CLI。
- soft-target dataset exporter。
- calibration step。
- model/evidence manifest。

**驗收**
- 同一 revision + seed 可重跑，metric drift 在預先定義 tolerance 內。
- trained candidate 必須優於自身 zero-shot baseline。
- **Intent promotion gate**：以「coverage at ≥85% graded accepted precision」作主要 metric；至少要明顯高於目前 intent n-gram 的低 coverage baseline，否則不值得進研究擴展。
- exact precision、graded precision、coverage、Brier/ECE、language slice 全部保存，不只報 accuracy。
- 若 fine-tune 仍無法在 precision/coverage tradeoff 上勝過 n-gram，先 STOP，不啟動 PRAXIST 大規模搜尋。

---

### Phase 4 — Context-aware state / ablation

**目的**

驗證 #233 已確認的規則：短回覆必須延續 work-in-progress；模型不能只看當前一句。

比較：
1. prompt-only
2. prompt + previous actionable request
3. prompt + structured active-task state

**文件**
- context ownership。
- 最大長度與 truncation。
- 不持久化 secrets / raw unrelated conversation。
- context unavailable 時的 fallback。

**RED tests**
- `keep going` / `do it` / `next one` 類 continuation。
- stale active task。
- conflicting previous/current instruction。
- oversized context。
- non-actionable prompt 不應被舊 context 強行變成 action。

**GREEN implementation**
- context serializer。
- ablation runner。
- evaluator slice：continuation vs standalone。

**驗收**
- continuation slice 必須有明顯改善。
- standalone slice 不得出現顯著 regression。
- context serialization deterministic。
- 無跨 task/session 的 stale context contamination。
- 若 structured context 沒有實質增益，保留較簡單的 prompt-only/previous-request 方案，不因架構偏好硬上。

---

### Phase 5 — PRAXIST task harness

**目的**

到這一步才把已能獨立跑的研究 project 交給 PRAXIST。PRAXIST **不得成為 evaluator owner，也不得進 production dependency**。

**task-owned objective**
至少包含：
- graded accepted precision ↑
- coverage ↑
- exact accepted precision ↑
- calibration error ↓
- continuation quality ↑
- run-to-run disagreement ↓
- training/inference resource usage ↓

**hard constraints**
- structural-floor violation = 0
- final-holdout leakage = 0
- public private-data leakage = 0
- evaluator modification = forbidden

**文件**
- candidate 可修改的檔案/參數範圍。
- metric direction。
- evidence maturity。
- compute / generation / peer budget。
- stop condition。
- promotion protocol。

**RED tests**
- peer 嘗試改 evaluator → fail。
- peer 嘗試把 final holdout 當 train/dev → fail。
- evidence 缺 provenance → fail。
- metric direction / baseline 缺失 → readiness fail。

**GREEN implementation**
- PRAXIST task harness。
- wrapper 呼叫同一個 #255 evaluator。
- evidence export 只含 hashes / aggregate metrics / approved artifacts。

**驗收**
- direct evaluator 與 PRAXIST wrapper 對同 checkpoint 得到一致 metrics。
- 先做 **1 generation × 2 peers** smoke run。
- run 可 stop / resume / replay。
- evidence chain 能回溯到 source revision、dataset revision、checkpoint hash。
- public Harness repo 不出現 private prompts。

---

### Phase 6 — Bounded research generations

**目的**

讓 PRAXIST 探索真正不確定的部分，而不是重新設計 specification。

第一輪 search lanes 建議限制為：
1. multilingual vs typed-decisions base。
2. prompt-only vs context-aware state。
3. per-intent `noul` wording / criteria。
4. calibration strategy。
5. joint intent + tier proposal 是否值得。
6. 只有前述有效後才探索 ONNX / quantization。

**禁止探索**
- 改 taxonomy 來讓分數變漂亮。
- 修改 final holdout。
- 移除 structural floor。
- 把 abstain/fallback 拿掉換 coverage。
- 直接最佳化 final holdout。

**研究方法**
- PRAXIST generations 只看 train/validation/research-dev evidence。
- final holdout 不可成為每輪 feedback signal。
- candidate freeze 後才允許 final-holdout promotion evaluation。

**驗收**
- 至少保留 baseline、frontier candidates、失敗候選與淘汰原因。
- 每個 promoted candidate 有可重播 config。
- candidate 的改善需跨 en / zh-TW 與 continuation/non-continuation slices，不接受只靠單 slice 拉平均。
- 若沒有 candidate 穩定超越 Phase 3 baseline，結論可以是 **不採 PRAXIST 找到的變體**；負結果仍視為成功研究輸出。

---

### Phase 7 — Runtime export / CPU feasibility

**目的**

研究結果只有在 #233 runtime gate 可接受時才有產品價值。

比較：
- PyTorch resident
- ONNX Runtime
- 若必要，再評估 quantization

**文件**
- export format。
- runtime dependencies。
- cold / warm lifecycle。
- model hash verification。
- offline behavior（production runtime 不得首次請求時下載模型）。

**RED tests**
- export score 與 training runtime 超 tolerance → fail。
- hash mismatch → fail。
- model unavailable/crash → fallback。
- first-use network dependency → fail。
- unsupported model revision → fail。

**GREEN implementation**
- exported candidate。
- resident provider integration prototype。
- benchmark runner。

**驗收**
- score parity 在定義 tolerance 內。
- 使用 #233 **既有 warm-latency gate**，不要另創較寬鬆標準。
- cold start、warm p50/p95、RSS、artifact size 都保存。
- Windows/Linux/macOS 至少 mechanism tests；實際 host performance 需標示測試平台。
- runtime 不依賴 PRAXIST。
- 若 quality 過關但 CPU gate 不過，結果標成 research-success / rollout-blocked，而不是偷偷放寬 gate。

---

### Phase 8 — Final promotion / handoff back to #233

**目的**

#255 不自行宣布 production replacement；最後把 frozen candidate 與完整 evidence 交還 #233。

**final evaluation**
- frozen model/checkpoint hash。
- frozen adapter revision。
- frozen evaluator revision。
- frozen final holdout。
- 一次正式 promotion run。

**驗收**
- #233 的 precision / coverage / determinism / latency / provenance / fallback gates 全部逐項列出 PASS/FAIL。
- structural-floor violation = **0**。
- final holdout 沒有在研究過程被拿來調參。
- artifact 可由 manifest 重建或驗 hash取得。
- 有 rollback/fallback 證據。
- 結論只能是：
  - **promotable**：建立 #233 integration PR / follow-up。
  - **not promotable**：保留 evidence，#233 繼續既有方案。
  - **blocked**：明確列出尚缺的 runtime/data/license 前置。

---

## Phase dependencies

```text
0 contract
  ↓
1 dense labels
  ↓
2 Laya zero-shot
  ↓
3 fine-tune baseline ───────┐
  ↓                        │
4 context ablation         │
  ↓                        │
5 PRAXIST harness          │
  ↓                        │
6 bounded research         │
  ↓                        │
7 CPU/runtime feasibility  │
  ↓                        │
8 #233 promotion decision ◄┘
```

## 重要 stop/go 原則

1. **Phase 3 不勝過 baseline → 不啟動昂貴研究。**
2. **Phase 6 沒有穩定 frontier improvement → 不因為用了 PRAXIST 就硬推 Laya。**
3. **Phase 7 quality 過但 CPU/runtime gate 不過 → 不進 production。**
4. **Phase 8 只有 evidence 能決定是否回到 #233 rollout；不能以架構偏好取代 gate。**

這樣 #255 的成功條件不是「一定把 Laya 上線」，而是 **對 Laya + PRAXIST 是否值得成為 #233 的下一代 System One 路徑，產生可重現、可審核、可停止的答案**。
